File: SimpleEngine/fse/Provider.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Xml;

namespace SimpleEngine.fse
{
    public abstract class Provider
    {
        Target target;
        readonly string targetname;

        private string id = String.Empty;

        public string Id
        {
            get
            {
                return id;
            }
            set
            {
                if (string.IsNullOrEmpty(id)) id = value;
                else throw new Exception("Unable to change id from " + id + " to " + value);
            }
        }

        public override string ToString()
        {
            return Id + "(" + targetname + "): <" + new StringSeperator(",", " and ", "").Append(commands.ToArray()).ToString() + ">";
        }

        private bool autocallCommands = true;

        public void provide(RenderArgs ra)
        {
            if (autocallCommands)
            {
                callCommands();
            }

            target.apply(delegate
            {
                doProvide(ra);
            });
        }


        protected void denyAutocallOfCommands()
        {
            autocallCommands = false;
        }

        protected void callCommands()
        {
            foreach (Command c in commands)
            {
                c.apply();
            }
        }
        protected abstract void doProvide(RenderArgs ra);

        internal Target Target
        {
            get
            {
                return target;
            }
        }

        public class Link
        {
            Provider prov = null;
            readonly string name;

            public Link(string name)
            {
                this.name = name;
            }

            public void provide(RenderArgs ra)
            {
                prov.doProvide(ra);
            }

            public void sortout(Linker usr)
            {
                prov = usr.getProvider(name);
            }
        }

        internal BufferReference createBuffer(string name)
        {
            return new BufferReference(name);
        }

        protected Provider(XmlElement el)
        {
            targetname = Xml.GetAttributeString(el, "target");

            foreach (XmlElement e in Xml.Elements(el))
            {
                commands.Add(Commands.Commands.Create(e, this));
            }
        }

        private List<Command> commands = new List<Command>();
        private List<Provider> providers = new List<Provider>();

        public IEnumerable<Provider> Providers
        {
            get
            {
                return providers;
            }
        }

        public void link(Linker linker)
        {
            if (false == string.IsNullOrEmpty(targetname))
            {
                target = linker.getTarget(targetname);
                target.Provider = this;
            }

            foreach (Command c in commands)
            {
                c.link(linker);
            }

            doLink(linker);
        }

        protected abstract void doLink(Linker linker);

        public void bind(Binder bd)
        {
            doBind(bd);

            foreach (Command c in commands)
            {
                c.bind(bd);
            }
        }
        protected abstract void doBind(Binder bd);

        internal void postlink(Linker linker)
        {
            foreach (Command c in commands)
            {
                c.link(linker);

                foreach (Provider p in c.Dependencies)
                {
                    if (p == null) throw new NullReferenceException();
                    this.providers.Add(p);
                }
            }
        }
    }
}

File: src/engine/fse/provider.cc
#include "engine/fse/provider.h"

#include <stdexcept>

#include "engine/fse/command.commands.h"
#include "engine/fse/command.h"
#include "engine/fse/linker.h"
#include "engine/fse/renderargs.h"
#include "engine/fse/target.h"
#include "engine/stringseperator.h"
#include "fmt/core.h"

namespace simple::fse
{
    const std::string& Provider::GetId() const
    {
        return id;
    }

    void Provider::OnSize(int width, int height)
    {
        target->OnSize(width, height);
    }

    Provider::Provider(std::shared_ptr<xml::Element> el)
        : target_name(xml::GetAttributeString(el, "target"))
        , id(xml::GetAttributeString(el, "id"))
    {
    }

    void Provider::PostLoad(std::shared_ptr<Provider> provider, std::shared_ptr<xml::Element> el)
    {
        for (std::shared_ptr<xml::Element> e : xml::Elements(el))
        {
            provider->commands.emplace_back(commands::Create(e, provider));
        }
    }

    Provider::~Provider()
    {
    }

    void Provider::SetId(const std::string& value)
    {
        if (id.empty())
        {
            id = value;
        }
        else
        {
            throw std::runtime_error(fmt::format("Unable to change id from {} to {}", id, value));
        }
    }

    std::string Provider::ToString() const
    {
        std::vector<std::string> command_names;
        for (auto& c : commands)
        {
            command_names.emplace_back(c->ToString());
        }
        const auto command_string = StringSeperator(",", " and ", "").ToString(command_names);
        return fmt::format("{}({}): <{}>", GetId(), target_name, command_string);
    }

    void Provider::Provide(RenderArgs* ra)
    {
        if (autocall_commands)
        {
            CallCommands();
        }

        target->Apply([this, ra]() { PostProvide(ra); });
    }

    void Provider::DenyAutocallOfCommands()
    {
        autocall_commands = false;
    }

    void Provider::CallCommands()
    {
        for (auto c : commands)
        {
            c->Apply();
        }
    }

    std::shared_ptr<BufferReference> Provider::CreateBuffer(const std::string& name)
    {
        return std::make_shared<BufferReference>(name);
    }

    void Provider::Link(std::shared_ptr<Provider> provider, Linker* linker)
    {
        if (provider->target_name != "")
        {
            provider->target = linker->GetTarget(provider->target_name);
            provider->target->SetProvider(provider);
        }

        for (auto c : provider->commands)
        {
            c->Link(linker);
        }

        provider->PostLink(linker);
    }

    void Provider::Bind(Binder* bd)
    {
        PreBind(bd);

        for (auto c : commands)
        {
            c->Bind(bd);
        }
    }

    void Provider::OnLinkCompleted(Linker* linker)
    {
        for (auto c : commands)
        {
            // why call link twice?
            c->Link(linker);

            for (auto p : c->GetDependencies())
            {
                if (p == nullptr)
                {
                    throw std::runtime_error("null pointer exception");
                }
                providers.emplace_back(p);
            }
        }
    }
}
